### DISCUSSION TITLE
Visualizing 3D label maps from medical image segmentation like itksnap

### DISCUSSION THREAD

**Discussion by jmargeta** [pyvista/pyvista#2434] [Q&A] [answered]
Hi PyVista community :wave: 

What is the best way to visualize 3D label maps as are often used in medical imaging to represent segmentation output or ground-truth?  The label map is a 3D volume where value of each voxel uniquely identifies the underlying structure of interest (e.g. in the example below, 1 is the left atrium, 2, 3, 4, 5 are pulmonary arteries).

The values are categorical (ints) but not necessarily contiguous, the value of the background is often 0.
The different labels can be nested within each other so would be valuable to be able to show/hide/set transparency of each individual structure as needed.

This is a simple solution I have for now:
```python
import pyvista as pv

label = pv.read("example_label.mha")
num_classes = 5
pl = pv.Plotter()
pl.add_mesh(
  label.contour(),
  opacity=(num_classes + 1) * [1],
  clim=[0, num_classes], 
  cmap=plt.cm.get_cmap("viridis", num_classes + 1)
)

pl.view_zy()
pl.show()
```

Using it, I observe that there is a thin hull (or halo effect) around each label (I presume this is from marching cubes generating multiple contours whenever a transition from 0 occurs).
![halo](https://user-images.githubusercontent.com/1239359/161841467-15d214aa-877c-4524-91f0-4599acfaa159.png)
Surprisingly all is transparent even with opacity set to 1.

Without setting the opacity the pulmonary all is hidden under one surface of the same color:
![single](https://user-images.githubusercontent.com/1239359/161841481-b4ab580d-35ce-41d4-afc3-85fa4b445953.png)

Is there a way to remove the hull and make the segments fully opaque like in ITKSnap for example?
![itksnap](https://user-images.githubusercontent.com/1239359/161840402-33d47ea5-31d9-4b2c-8ba0-de5f202f3b7e.png)

What would be the best way to show only some labels and ignore the rest and not have any of those hulls?

Thresholding the label map per individually per class might work but becomes fairly slow when the number of classes grows. 

[example_label.zip adapted from the Medical decathlon (http://medicaldecathlon.com/) CC-SA](https://github.com/pyvista/pyvista/files/8421401/example_label.zip)

**Comment by njneeteson** [comment, marked as answer]
Have you tried `image_threshold` or only tried `threshold`? I think the former should be a lot faster than the latter since it won't change the mesh topology and it will allow you to generate a binary image for each label that you can then visualize separately however you like.

**Reply by jmargeta** [reply]
Wonderful, thanks @njneeteson for the pointer! ```image_threshold``` is a clearly much faster way to select each label individually for this usecase.
